### ISSUE TITLE
Unable to install (Error: 'cstring' file not found) - MacOS Mojave

### ISSUE THREAD

**Issue by charlieyou415** [sirfz/tesserocr#162]
```
Collecting tesserocr
  Using cached https://files.pythonhosted.org/packages/f8/6d/4e81e041f33a4419e59edcb1dbdf3c56e9393f60f5ef531381bd67a1339b/tesserocr-2.3.1.tar.gz
Building wheels for collected packages: tesserocr
  Running setup.py bdist_wheel for tesserocr ... error
  Complete output from command /Users/charlieyou/miniconda3/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/2h/ty244jbd1vd3d1l0m4k3_g980000gn/T/pip-install-wgh41bv0/tesserocr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /private/var/folders/2h/ty244jbd1vd3d1l0m4k3_g980000gn/T/pip-wheel-5b3wyn1i --python-tag cp36:
  pkg-config failed to find tesseract/lept libraries: b"Package tesseract was not found in the pkg-config search path.\nPerhaps you should add the directory containing `tesseract.pc'\nto the PKG_CONFIG_PATH environment variable\nNo package 'tesseract' found\n"
  Supporting tesseract v4.0.0
  Building with configs: {'libraries': ['tesseract', 'lept'], 'cython_compile_time_env': {'TESSERACT_VERSION': 262144}}
  running bdist_wheel
  running build
  running build_ext
  building 'tesserocr' extension
  creating build
  creating build/temp.macosx-10.7-x86_64-3.6
  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/charlieyou/miniconda3/include -arch x86_64 -I/Users/charlieyou/miniconda3/include -arch x86_64 -I/usr/local/include -I/Users/charlieyou/miniconda3/include/python3.6m -c tesserocr.cpp -o build/temp.macosx-10.7-x86_64-3.6/tesserocr.o -std=c++11 -DUSE_STD_NAMESPACE
  warning: include path for stdlibc++ headers not found; pass '-std=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
  In file included from tesserocr.cpp:603:
  **/usr/local/include/tesseract/unichar.h:24:10: fatal error: 'cstring' file not found
  #include <cstring>**
           ^~~~~~~~~
  1 warning and 1 error generated.
  error: command 'gcc' failed with exit status 1

  ----------------------------------------
  Failed building wheel for tesserocr
  Running setup.py clean for tesserocr
Failed to build tesserocr
Installing collected packages: tesserocr
  Running setup.py install for tesserocr ... error
    Complete output from command /Users/charlieyou/miniconda3/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/2h/ty244jbd1vd3d1l0m4k3_g980000gn/T/pip-install-wgh41bv0/tesserocr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/2h/ty244jbd1vd3d1l0m4k3_g980000gn/T/pip-record-0grp17wc/install-record.txt --single-version-externally-managed --compile:
    pkg-config failed to find tesseract/lept libraries: b"Package tesseract was not found in the pkg-config search path.\nPerhaps you should add the directory containing `tesseract.pc'\nto the PKG_CONFIG_PATH environment variable\nNo package 'tesseract' found\n"
    Supporting tesseract v4.0.0
    Building with configs: {'libraries': ['tesseract', 'lept'], 'cython_compile_time_env': {'TESSERACT_VERSION': 262144}}
    running install
    running build
    running build_ext
    building 'tesserocr' extension
    creating build
    creating build/temp.macosx-10.7-x86_64-3.6
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/charlieyou/miniconda3/include -arch x86_64 -I/Users/charlieyou/miniconda3/include -arch x86_64 -I/usr/local/include -I/Users/charlieyou/miniconda3/include/python3.6m -c tesserocr.cpp -o build/temp.macosx-10.7-x86_64-3.6/tesserocr.o -std=c++11 -DUSE_STD_NAMESPACE
    warning: include path for stdlibc++ headers not found; pass '-std=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
    In file included from tesserocr.cpp:603:
    **/usr/local/include/tesseract/unichar.h:24:10: fatal error: 'cstring' file not found
    #include <cstring>**
             ^~~~~~~~~
    1 warning and 1 error generated.
    error: command 'gcc' failed with exit status 1

    ----------------------------------------
Command "/Users/charlieyou/miniconda3/bin/python -u -c "import setuptools, tokenize;__file__='/private/var/folders/2h/ty244jbd1vd3d1l0m4k3_g980000gn/T/pip-install-wgh41bv0/tesserocr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/2h/ty244jbd1vd3d1l0m4k3_g980000gn/T/pip-record-0grp17wc/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /private/var/folders/2h/ty244jbd1vd3d1l0m4k3_g980000gn/T/pip-install-wgh41bv0/tesserocr/
```

Hi, 

I'm trying to install tesserocr but encountering the above error. 
I've tried a few different ways to install, as well as a different potential fixes in previous posts, all of which with no avail. 

At this point I'm certain the issue is not related to this package (since my friend installed it with no problems, I also installed on a Ubuntu VM w/o issue), however I'm still posting here hoping for people who are close to this package can offer some insights. 

My tesseract (installed through brew) version is: 
`tesseract 4.0.0
 leptonica-1.76.0
  libjpeg 9c : libpng 1.6.35 : libtiff 4.0.9 : zlib 1.2.11
`

Thanks and let me know what other info you might need.

**Comment by VeneetReddyK** [comment]
+1

**Comment by Yang-Programmer** [comment]
I have the same problem
OS ：macOS Mojave  10.14.1
tesseract : 4.0.0

**Comment by sirfz** [comment]
cstring is a c++ header file which seems to be missing from your system.

**Comment by zmjios** [comment]
I  also have the same problem
OS ：macOS Mojave 10.14.1
tesseract : 4.0.0

**Comment by zmjios** [comment]
Installing collected packages: tesserocr
  Running setup.py install for tesserocr ... error
    Complete output from command /Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 -u -c "import setuptools, tokenize;__file__='/private/var/folders/pt/9nxbnsp14278rpqr6bkk475r0000gp/T/pip-install-gqwjw1bo/tesserocr/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/var/folders/pt/9nxbnsp14278rpqr6bkk475r0000gp/T/pip-record-m6mt4j7y/install-record.txt --single-version-externally-managed --compile:
    Supporting tesseract v4.0.0
    Configs from pkg-config: {'include_dirs': ['/usr/local/Cellar/leptonica/1.76.0/include', '/usr/local/Cellar/tesseract/4.0.0/include'], 'library_dirs': ['/usr/local/Cellar/tesseract/4.0.0/lib', '/usr/local/Cellar/leptonica/1.76.0/lib'], 'libraries': ['tesseract', 'lept'], 'cython_compile_time_env': {'TESSERACT_VERSION': 262144}}
    /Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/distutils/dist.py:274: UserWarning: Unknown distribution option: 'long_description_content_type'
      warnings.warn(msg)
    running install
    running build
    running build_ext
    building 'tesserocr' extension
    creating build
    creating build/temp.macosx-10.9-x86_64-3.7
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch x86_64 -g -I/usr/local/Cellar/leptonica/1.76.0/include -I/usr/local/Cellar/tesseract/4.0.0/include -I/Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m -c tesserocr.cpp -o build/temp.macosx-10.9-x86_64-3.7/tesserocr.o -std=c++11 -DUSE_STD_NAMESPACE
    In file included from tesserocr.cpp:47:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/Python.h:30:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string.h:61:
    In file included from /usr/local/include/string.h:25:
    In file included from /usr/local/include/plist/Node.h:25:
    /usr/local/include/plist/plist.h:161:28: error: unknown type name 'uint64_t'
        plist_t plist_new_uint(uint64_t val);
                               ^
    /usr/local/include/plist/plist.h:180:45: error: unknown type name 'uint64_t'
        plist_t plist_new_data(const char *val, uint64_t length);
                                                ^
    /usr/local/include/plist/plist.h:199:27: error: unknown type name 'uint64_t'
        plist_t plist_new_uid(uint64_t val);
                              ^
    /usr/local/include/plist/plist.h:437:43: error: unknown type name 'uint64_t'
        void plist_get_uint_val(plist_t node, uint64_t * val);
                                              ^
    /usr/local/include/plist/plist.h:457:55: error: unknown type name 'uint64_t'
        void plist_get_data_val(plist_t node, char **val, uint64_t * length);
                                                          ^
    /usr/local/include/plist/plist.h:476:42: error: unknown type name 'uint64_t'
        void plist_get_uid_val(plist_t node, uint64_t * val);
                                             ^
    /usr/local/include/plist/plist.h:520:43: error: unknown type name 'uint64_t'
        void plist_set_uint_val(plist_t node, uint64_t val);
                                              ^
    /usr/local/include/plist/plist.h:540:60: error: unknown type name 'uint64_t'
        void plist_set_data_val(plist_t node, const char *val, uint64_t length);
                                                               ^
    /usr/local/include/plist/plist.h:559:42: error: unknown type name 'uint64_t'
        void plist_set_uid_val(plist_t node, uint64_t val);
                                             ^
    In file included from tesserocr.cpp:47:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/Python.h:30:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string.h:61:
    In file included from /usr/local/include/string.h:26:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string:477:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string_view:176:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__string:56:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:641:
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:70:9: error: no member named 'memcpy' in the global namespace
    using ::memcpy;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:71:9: error: no member named 'memmove' in the global namespace; did you mean 'remove'?
    using ::memmove;
          ~~^
    /usr/include/stdio.h:174:6: note: 'remove' declared here
    int      remove(const char *);
             ^
    In file included from tesserocr.cpp:47:
    In file included from /Library/Frameworks/Python.framework/Versions/3.7/include/python3.7m/Python.h:30:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string.h:61:
    In file included from /usr/local/include/string.h:26:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string:477:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string_view:176:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__string:56:
    In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:641:
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:72:9: error: no member named 'strcpy' in the global namespace
    using ::strcpy;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:73:9: error: no member named 'strncpy' in the global namespace
    using ::strncpy;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:74:9: error: no member named 'strcat' in the global namespace
    using ::strcat;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:75:9: error: no member named 'strncat' in the global namespace
    using ::strncat;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:76:9: error: no member named 'memcmp' in the global namespace
    using ::memcmp;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:77:9: error: no member named 'strcmp' in the global namespace
    using ::strcmp;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:78:9: error: no member named 'strncmp' in the global namespace
    using ::strncmp;
          ~~^
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/cstring:79:9: error: no member named 'strcoll' in the global namespace
    using ::strcoll;
          ~~^
    fatal error: too many errors emitted, stopping now [-ferror-limit=]
    20 errors generated.
    error: command 'gcc' failed with exit status 1

**Comment by PinkFloyded** [comment]
I was able to get it working with the following diff:

```diff
diff --git a/setup.py b/setup.py
index 81f5b26..95b86dd 100644
--- a/setup.py
+++ b/setup.py
@@ -134,7 +134,7 @@ def get_build_args():

     if build_args['cython_compile_time_env']['TESSERACT_VERSION'] >= 0x030502:
         _LOGGER.debug('tesseract >= 03.05.02 requires c++11 compiler support')
-        build_args['extra_compile_args'] = ['-std=c++11', '-DUSE_STD_NAMESPACE']
+        build_args['extra_compile_args'] = ['-std=c++11', "-stdlib=libc++", "-mmacosx-version-min=10.9", '-DUSE_STD_NAMESPACE']

     _LOGGER.debug('build parameters: {}'.format(build_args))
     return build_args
```

More info is in [this issue](https://github.com/pandas-dev/pandas/issues/23424) and specifically [this comment](https://github.com/pandas-dev/pandas/issues/23424#issuecomment-446393981).

**Comment by sirfz** [comment]
Hi @PinkFloyded, if this is a standard fix for MacOS then a PR targeting this system is welcome.

**Comment by poke1024** [comment]
@sirfz @PinkFloyded Seems to be a general issue at the moment, see https://github.com/cython/cython/issues/2694

**Comment by poke1024** [comment]
@PinkFloyded's fix works for me on macOS 10.14.5 / tesseract 4, but I also need to set `MACOSX_DEPLOYMENT_TARGET`, otherwise I get:

```
clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
ld: library not found for -lstdc++
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

So, this works for me:

```
export CFLAGS="-stdlib=libc++"
export MACOSX_DEPLOYMENT_TARGET=10.9
python setup.py install
```

**Comment by Zognge** [comment]
![image](https://user-images.githubusercontent.com/32995857/77710580-6b927b80-7009-11ea-902d-a98f2d88e064.png)
QT project also do.
